File: Sources/cpp/src/sprites/timed_content_provider.cpp
#include "timed_content_provider.h"

#include <chrono>
#include <iostream>
#include <vector>

template<typename T>
TimedContentProvider<T>::TimedContentProvider(const std::vector<T> frames, double fps)
    : frames(frames),
      currentFrameIndex(0),
      completedLoops(0),
      lastUpdateTime(0),
      lastFrameChangeTime(0)
{
    auto frameDurationMs = fps > 0.0f ? 1000.0 / fps : 0;
    frameDuration = std::chrono::milliseconds(uint32_t(frameDurationMs));
}

template<typename T>
const uint32_t TimedContentProvider<T>::numberOfFrames() const {
    return frames.size();
}

template<typename T>
const T& TimedContentProvider<T>::currentFrame() const {
    return frames[currentFrameIndex];
}

template<typename T>
void TimedContentProvider<T>::update(std::chrono::milliseconds timeSinceLastUpdate) {
    lastUpdateTime += timeSinceLastUpdate;

    if ((lastUpdateTime - lastFrameChangeTime) >= frameDuration) {
        loadNextFrame();
        lastFrameChangeTime = lastUpdateTime;
    }
}

template<typename T>
void TimedContentProvider<T>::loadNextFrame() {
    auto original = currentFrameIndex;
    uint32_t nextIndex = (currentFrameIndex + 1) % frames.size();
    checkLoopCompletion(nextIndex);
    currentFrameIndex = nextIndex;
    auto frame = currentFrame();
}

template<typename T>
void TimedContentProvider<T>::checkLoopCompletion(uint32_t nextIndex) {
    if (nextIndex < currentFrameIndex) {
        completedLoops++;
    }
}

template class TimedContentProvider<uint32_t>;
template class TimedContentProvider<std::string>;

File: Sources/windows/App/Sources/Sprites/TimedContentProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace App
{
    public class TimedContentProvider<T>
    {
        private List<T> frames;
        private TimeSpan frameDuration;
        private uint currentFrameIndex;
        private long completedLoops;
        private TimeSpan lastUpdateTime;
        private TimeSpan lastFrameChangeTime;

        public TimedContentProvider(List<T> frames, double fps)
        {
            this.frames = frames;
            currentFrameIndex = 0;
            completedLoops = 0;
            lastUpdateTime = TimeSpan.Zero;
            lastFrameChangeTime = TimeSpan.Zero;

            var frameDurationMs = fps > 0.0 ? 1000.0 / fps : 0;
            frameDuration = TimeSpan.FromMilliseconds(frameDurationMs);
        }

        public T CurrentFrame
        {
            get { return frames[(int)currentFrameIndex]; }
        }

        public void Update(TimeSpan timeSinceLastUpdate)
        {
            lastUpdateTime += timeSinceLastUpdate;

            if ((lastUpdateTime - lastFrameChangeTime) >= frameDuration)
            {
                LoadNextFrame();
                lastFrameChangeTime = lastUpdateTime;
            }
        }

        public uint NumberOfFrames
        {
            get { return (uint)frames.Count; }
        }

        private void LoadNextFrame()
        {
            var original = currentFrameIndex;
            uint nextIndex = (currentFrameIndex + 1) % (uint)frames.Count;
            CheckLoopCompletion(nextIndex);
            currentFrameIndex = nextIndex;
            var frame = CurrentFrame;
        }

        private void CheckLoopCompletion(uint nextIndex)
        {
            if (nextIndex < currentFrameIndex)
            {
                completedLoops++;
            }
        }
    }
}
